### PR TITLE
Add MediaQuery.systemGestureInsets to support Android Q

### PR DESCRIPTION
**Pull request by shihaohong**
## Description:

This passes gestural navigation inset information to the `MediaQuery.systemGestureInsets`. This becomes necessary as Flutter app developers want to take advantage of their entire screen with full gestural navigation mode coming with Q. This allows Flutter developers to determine the minimum distance from each edge needed for any swipe-able widgets like sliders.

## Related Issues

Related to https://github.com/flutter/flutter/issues/32748, https://github.com/flutter/flutter/issues/34678
Depends on https://github.com/flutter/engine/pull/10413

## Tests

I added the following tests:

- Updated existing tests such as `copyWith` and defaulting to source.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by shihaohong** [comment]
I'm currently working on the engine counterpart of this PR, so until that lands, this PR will remain a work in progress!

**Comment by shihaohong** [comment]
I'm closing this for now, since this is not actionable given that we're blocked on the engine PR. I will create a new PR once the engine PR is completed.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart]: "contains" is kinds strange. Isn't this just the area where the system may detect certain gestures, handle them internally, and not forward those to your app?

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart]: Can you add a little more information here, e.g. that certain gestures in these areas my  not be forwarded to the app and elements placed here may therefore not respond to them.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/media_query.dart]: I will update this to
```
  /// The parts of the display where the system may be detecting particular
  /// gestures.
  ///
  /// This value is typically a drag gesture from an edge of the screen.
```

Let me know if there is any additional improvement/clarification we could make

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/media_query.dart]: Sounds good, I did the following: 
```
  /// This value is typically a drag gesture from an edge of the screen.
  /// Gestures in these areas may not be forwarded to the app or may
  /// be canceled by the system. Therefore, elements placed in these areas may
  /// not properly respond to them.
```

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart]: Maybe:

The parts of the display where the system may be detecting certain gestures for system actions.

Those gestures are typically drags from the edge of the screen to perform system actions like opening the app drawer or the notification tray. Gestures reserved for system actions may not be forwarded to the app when performed in the area specified by these insets . Therefore, elements placed in these areas may not properly respond to them.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart]: We should say clearly on what systems this is used (e.g. Android starting with version Q)?

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/media_query.dart]: It's important to make this extra clear and part of that will be explaining exactly what AndroidQ does. So:

The areas along the edges of the display where the system consumes certain input gestures and blocks delivery to the app.

Starting with Android Q, simple swipe gestures are used by the system for page navigation and are not delivered to the app. Taps and swipe gestures that begin with a long-press _are_ delivered to the app but simple press-drag-release swipe gestures which begin within the area defined by [systemGestureInsets] are not.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/media_query.dart]: That's a great idea, I just added it.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/media_query.dart]: I tweaked the example a little (replaced the app drawer with back navigation), but incorporated everything else in your suggestion

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/media_query.dart]: You might be able to reuse a version of https://github.com/flutter/flutter/pull/37416#discussion_r316366337 here.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: For apps that might be deployed on Android Q devices with full ...

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: for the title, maybe 'Pad Slider to avoid systemGestureInsets'

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: Need a comment here to explain that, by default, the slider expands to fill the available width.

Not clear why we're padding top/bottom (we're assuming something about systemGestureInsets?).  Maybe: 
```
Container(
  alignment: Alignment.center,
  padding: systemGestureInsets, // only left,right padding is likely to be needed
  child: Slider...
),
```

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/media_query.dart]: There's a little nuance to this when I was testing it yesterday... The swipe gesture can still be registered by the app. I've gotten a drawer to open on a simple swipe, but it simultaneously triggers the back navigation arrow to appear. However, this is not consistent and sometimes, the drawer will appear and almost immediately disappear once the back navigation arrow starts to appear. 

I'll try to share a gif, but I think the general idea is that there will be conflicts in how simple drag gestures are handled in these inset areas and that devs should just avoid it anyway.

File: packages/flutter/test/widgets/media_query_test.dart
Comment (HansMuller) [inline packages/flutter/test/widgets/media_query_test.dart]: Separate PR: define these double values with a consts and explain why we're using strange values.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/media_query.dart]: Did you have a chance to just place a Listener [1] in this area and see what events you actually get in Flutter?

[1] https://master-api.flutter.dev/flutter/widgets/Listener-class.html

Comment (shihaohong) [reply packages/flutter/lib/src/widgets/media_query.dart]: All this to say, maybe updating the wording to "simple swipe gestures that start within the [systemGestureInsets] areas are used by the system for page navigation and ~are~ may not *be* delivered to the app"

Comment (HansMuller) [reply packages/flutter/lib/src/widgets/media_query.dart]: Sounds OK to me. We can tighten up the wording in the future, if we need to.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/media_query.dart]: I mainly pad the top/bottom to make it look cleaner in the sample app. I'll just set top and bottom to 0 for the sake of the sample to avoid confusion, which seems like a worthy tradeoff

File: packages/flutter/test/widgets/media_query_test.dart
Comment (shihaohong) [reply packages/flutter/test/widgets/media_query_test.dart]: Filed https://github.com/flutter/flutter/issues/39012

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: In this case you could use `EdgeInsets.only`

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: To avoid having the left and right edges of the slider appear within the area reserved for system gesture navigation.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: may not => may not be

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: [my fault] "There is no reason" is probably overstating it. Maybe: Apps should feel free to put visual elements within

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/media_query.dart]: @goderbauer I just played with the `Listener`, and it seems like it lets all gestures through and if you start doing a horizontal swipe, it sends a cancel event and begins the system gesture. 

Oddly, if you do a diagonal swipe from the edge (ie. top-left to bottom-right or bottom-left to top-right), that seems to never trigger the system gesture. I decided to try this after reading about it as a "workaround" on some articles on the internet.

**Pull request merged**
